File: src/main/javaCode/com/hjf/java/spark/Test01.java
package com.hjf.java.spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.*;
import scala.Tuple2;
import scala.tools.nsc.util.ClassPath;

import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

/**
 * @author Jiang锋时刻
 * @create 2020-04-03 0:51
 */
public class Test01 {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setAppName("Test01");
        conf.setMaster("local");

        JavaSparkContext sc = new JavaSparkContext(conf);
        JavaRDD<String> lines = sc.textFile("./data/words.txt");

//        JavaRDD<String> result = lines.filter(new Function<String, Boolean>() {
//            @Override
//            public Boolean call(String line) throws Exception {
//                return "hello world".equals(line);
//            }
//        });

//        result.foreach(new VoidFunction<String>() {
//            @Override
//            public void call(String s) throws Exception {
//                System.out.println(s);
//            }
//        });

//        System.out.println(result.count());



//        JavaRDD<Object> map = lines.map(new Function<String, Object>() {
//            @Override
//            public Object call(String s) throws Exception {
//                return s + "*";
//            }
//        });
//
//        map.foreach(new VoidFunction<Object>() {
//            @Override
//            public void call(Object o) throws Exception {
//                System.out.println(o);
//            }
//        });



//        JavaPairRDD<String, String> result = lines.mapToPair(new PairFunction<String, String, String>() {
//
//            @Override
//            public Tuple2<String, String> call(String s) throws Exception {
//                return new Tuple2<>(s, s + "#");
//            }
//        });
//
//        result.foreach(new VoidFunction<Tuple2<String, String>>() {
//            @Override
//            public void call(Tuple2<String, String> stringStringTuple2) throws Exception {
//                System.out.println(stringStringTuple2);
//            }
//        });


//        JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
//            @Override
//            public Iterator<String> call(String s) throws Exception {
//                List<String> list = Arrays.asList(s.split(" "));
//                return list.iterator();
//
//            }
//        });
//
//        JavaPairRDD<String, Integer> pairWords = words.mapToPair(new PairFunction<String, String, Integer>() {
//            @Override
//            public Tuple2<String, Integer> call(String word) throws Exception {
//                return new Tuple2<>(word, 1);
//            }
//        });
//
//        JavaPairRDD<String, Integer> resultRDD = pairWords.reduceByKey(new Function2<Integer, Integer, Integer>() {
//            @Override
//            public Integer call(Integer v1, Integer v2) throws Exception {
//                return v1 + v2;
//            }
//        });
//
//        // key value 交换顺序
//        JavaPairRDD<Integer, String> swapRDD = resultRDD.mapToPair(new PairFunction<Tuple2<String, Integer>, Integer, String>() {
//            @Override
//            public Tuple2<Integer, String> call(Tuple2<String, Integer> stringIntegerTuple2) throws Exception {
//                return stringIntegerTuple2.swap();
//            }
//        });
//
//        //排序
//        JavaPairRDD<Integer, String> sortRDD = swapRDD.sortByKey(false);
//
//        // 交换回来
//        JavaPairRDD<String, Integer> result = sortRDD.mapToPair(new PairFunction<Tuple2<Integer, String>, String, Integer>() {
//            @Override
//            public Tuple2<String, Integer> call(Tuple2<Integer, String> integerStringTuple2) throws Exception {
//                return integerStringTuple2.swap();
//            }
//        });
//
//
//        result.foreach(new VoidFunction<Tuple2<String, Integer>>() {
//            @Override
//            public void call(Tuple2<String, Integer> tp) throws Exception {
//                System.out.println(tp);
//            }
//        });


//        List<String> collect = lines.collect();
//
//        for (String one: collect){
//            System.out.println(one);
//        }


        JavaRDD<String> sample = lines.sample(true, 0.1);
        sample.foreach(new VoidFunction<String>() {
            @Override
            public void call(String s) throws Exception {
                System.out.println(s);
            }
        });


        sc.stop();
    }
}

File: src/main/scalaCode/com/hjf/spark/Test01.scala
package com.hjf.spark

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

/**
 * @author Jiang锋时刻
 * @create 2020-04-02 23:22
 */
object Test01 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
    conf.setAppName("Test01")
    conf.setMaster("local")

    val sc = new SparkContext(conf)

//    val rdd: RDD[String] = sc.parallelize(Array[String]("a", "b"))
    val rdd: RDD[String] = sc.makeRDD(Array[String]("a", "b"))
    rdd.foreach(println)



//    val lines: RDD[String] = sc.textFile("./data/words.txt")

    // map:一对一
    // lines.map(one => {one + "#"}).foreach(println)

    // flatMap:一对多
    // lines.flatMap(one => {one.split(" ")}).foreach(println)

    // filter:过滤
    // val words: RDD[String] = lines.flatMap(one => {one.split(" ")})
    // words.filter(one => {"hello".equals(one)}).foreach(println)

//    val words: RDD[String] = lines.flatMap(one => {
//      one.split(" ")
//    })
//    val pairWords: RDD[(String, Int)] = words.map(one => (one, 1))
//
//    val reduce: RDD[(String, Int)] = pairWords.reduceByKey((v1: Int, v2: Int) => {v1 + v2})

    // sortBy:排序.默认正序. true:正序, false:倒序
    // val result: RDD[(String, Int)] = reduce.sortBy(tp => tp._2, false)
    // result.foreach(println)

    // val transReduce: RDD[(Int, String)] = reduce.map(tp => {tp.swap})
    // val result: RDD[(Int, String)] = transReduce.sortByKey(false)
    // result.map(_.swap).foreach(println)

    // sample():
    // 第一个参数,是否有放回的抽样
    // 第二个参数,设置抽取比例 [接近这个比例]
    // 第三个参数,设置种子.设置后,重复执行时,结果不会发生变化
    // val result: RDD[String] = lines.sample(true, 0.1, 100L)
    // result.foreach(println)

    // count()
    // val count: Long = lines.count()
    // println(count)

    // collect():将结果回收到Driver端
    // val result: Array[String] = lines.collect()
    // result.foreach(println)

    // first():获取第一条数据
    // val result: String = lines.first()
    // println(result)

    //take(n):获取前n条数据
//    val result: Array[String] = lines.take(5)
//    result.foreach(println)



  }

}
